parallelize feed polling

saurfang opened the pull request:
Based on profiling, almost the entire delay on initial load of prove screen is attributed to subscription feed polling during state initialization. This PR simply parallelizes the polling to cut the loading time in half (there are two feeds we poll today.) 

Also, the endpoint itself is pretty slow taking 100ms~1s and most of the time is spend on semaphore proof verification in the `checkUserExists`. 
 https://github.com/proofcarryingdata/zupass/blob/883e50956033f4bdba910f6b5ba662c56e1cda41/apps/passport-server/src/services/issuanceService.ts#L424

```
passport-server:dev: checkUserExists1694834530611: 0.911ms deserialized signature
passport-server:dev: checkUserExists1694834530611: 580.097ms verified signature
passport-server:dev: checkUserExists1694834530611: 580.746ms fetched commitment from db
passport-server:dev: checkUserExists1694834530611: 580.769ms
```

I am not too familiar with the security assumption here but my guess for the long term solutions could be a combination of the following:
1. Poll subscription feed and update UI async instead of blocking the initial rendering
2. Can we cache the verification function in memory or in db (assume we can't speed up verification)
3. Can we exchange some short lived auth token instead (especially for read operations)

<img width="422" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/4ef0d5be-0268-4edd-94c5-09c1179eb7a9">


relates to https://github.com/proofcarryingdata/zupass/issues/597

robknight commented:
@ichub I think we could make this asynchronous. The blocking comes from the proving screen using `useIsSyncSettled`, which will only return true after feeds have been polled. We could either stop using this hook, or refactor the `sync` function and associated hooks (`useLoadedIssuedPCDs`) so that we're not relying on all subscriptions having completed polling.

Also at some point we are going to have third party feeds, which means that the risk of this process taking a long time both a) goes up and b) moves out of our control.

ichub commented on packages/passport-interface/src/SubscriptionManager.ts:
```diff
@@ -55,25 +55,26 @@ export class FeedSubscriptionManager {
-        console.log(`failed to poll subscription`, e);
-      }
-    }
+    const results = await Promise.all(
```
if any feed fails here, that will cause `Promise.all` to reject, which is not the behavior we want. You should instead use `Promise.allSettled`, and then go through each result, and react to each promise result appropriately.

ichub commented:
@saurfang,

> Poll subscription feed and update UI async instead of blocking the initial rendering
> Can we cache the verification function in memory or in db (assume we can't speed up verification)
> Can we exchange some short lived auth token instead (especially for read operations)

all reasonable suggestions. i think i would prefer to hold off on all of them for now, but if I had to choose any of them to implement, i would choose number 2. perhaps a fixed-size LRU cache service with some fixed size could be useful to have on the backend.

@robknight,

> @ichub I think we could make this asynchronous. The blocking comes from the proving screen using useIsSyncSettled, which will only return true after feeds have been polled. We could either stop using this hook here, or refactor the sync function and associated hooks (useLoadedIssuedPCDs) so that we're not relying on all subscriptions having completed polling.
>
> Also at some point we are going to have third party feeds, which means that the risk of this process taking a long time both a) goes up and b) moves out of our control.

the main reason we're using this hook in the proof screens is that we want to be super sure that the ticket PCDs are available to use there. I think we could create a configuration option for feeds - some of them block the rendering of the prove screens, and some of them (like frogs) don't. I don't feel strongly about doing this right now.

saurfang commented:
Makes sense and thanks for explaining. I might in a separate PR look into if it is easy to make this non-blocking and render a skeleton UI which can always resolves to an error state if no valid PCDs are found.

Addressed the suggestions using `allSettled` to guard against errors. ptal? @ichub

robknight commented:
This is getting substantially refactored in the subscription UI PR with a new system for tracking errors when fetching feeds, so I would prefer if we don't merge this change in. The catch on the performance implications is very much appreciated though!

saurfang commented:
Sounds good! I'll let you take it from here.
